sdks/java/io/distributedlog/src/main/java/org/apache/beam/sdk/io/distributedlog/DLBoundedSource.java

```diff
@@ -0,0 +1,192 @@
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import com.twitter.distributedlog.DistributedLogConfiguration;
+import com.twitter.distributedlog.DistributedLogConstants;
+import com.twitter.distributedlog.DistributedLogManager;
+import com.twitter.distributedlog.LogSegmentMetadata;
+import com.twitter.distributedlog.namespace.DistributedLogNamespace;
+import com.twitter.distributedlog.namespace.DistributedLogNamespaceBuilder;
+import java.io.IOException;
+import java.net.InetAddress;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.bookkeeper.stats.NullStatsLogger;
+import org.apache.commons.configuration.ConfigurationException;
+
+/**
+ * A {@code BoundedSource} for reading log streams resident in a DistributedLog cluster.
+ */
+class DLBoundedSource<R> extends BoundedSource<R> {
+
+  private static final long serialVersionUID = 3551126210922557330L;
+
+  protected final URI dlUri;
+  protected final Optional<URI> confUri;
+  protected final List<String> streams;
+  protected final Coder<R> rCoder;
+  protected final @Nullable LogSegmentBundle segment;
+
+  DLBoundedSource(
+      URI dlUri,
+      Optional<URI> confUri,
+      List<String> streams,
+      Coder<R> rCoder,
+      LogSegmentBundle segment) {
+    this.dlUri = dlUri;
+    this.confUri = confUri;
+    this.streams = streams;
+    this.rCoder = rCoder;
+    this.segment = segment;
+  }
+
+  DistributedLogNamespace setupNamespace() throws IOException {
+    DistributedLogConfiguration conf = new DistributedLogConfiguration();
+    if (confUri.isPresent()) {
+      try {
+        conf.loadConf(confUri.get().toURL());
+      } catch (ConfigurationException e) {
+        throw new IOException("Failed to load distributedlog configuration from " + confUri, e);
+      } catch (MalformedURLException e) {
+        throw new IOException("Invalid distributedlog configuration uri " + confUri, e);
+      }
+    }
+    return DistributedLogNamespaceBuilder.newBuilder()
+        .conf(conf)
+        .uri(dlUri)
+        .clientId("distributedlog-bounded-source-" + InetAddress.getLocalHost().getHostName())
+        .statsLogger(NullStatsLogger.INSTANCE)
+        .regionId(DistributedLogConstants.LOCAL_REGION_ID)
+        .build();
+  }
+
+  List<String> getStreams() {
+    return streams;
+  }
+
+  @Override
+  public void validate() {
+    checkArgument(!streams.isEmpty(), "need to set the streams of a DLBoundedSource");
+    checkNotNull(rCoder, "need to set the key coder of a DLBoundedSource");
```

Comment:
Ah, I need to change the logging. 'R' is the a record. I will add more documentation for this.